src/core/storage/sframe_interface/unity_sframe.cpp

```diff
@@ -1200,6 +1202,72 @@ unity_sframe::random_split(float percent, uint64_t random_seed, bool exact) {
   return logical_filter_split(logical_filter_array);
 }
 
+std::shared_ptr<unity_sframe_base> unity_sframe::shuffle() {
+  log_func_entry();
+
+  std::vector<std::string> column_names = this->column_names();
+  const size_t num_buckets = (this->size() / SFRAME_SHUFFLE_BUCKET_SIZE) + 1;
+
+  // Create a column of random ints between 0 and (num_buckets - 1).
+  auto temp_groupby_column = std::static_pointer_cast<unity_sarray>(
+    unity_sarray::make_uniform_int_array(this->size(), num_buckets));
+  const std::string rand_int_column_name("Random Ints");
+  std::shared_ptr<unity_sframe> temp(new unity_sframe());
+  temp->add_column(temp_groupby_column, rand_int_column_name);
+
+  // Pack columns so we can group by concatenate
+  std::shared_ptr<unity_sarray_base> packed_columns = this->pack_columns(column_names,
```

Comment:
great! Reduces overall random read times from the original shuffle index then read from index approach!